### DISCUSSION TITLE
Camera Controls and Screenshots

### DISCUSSION THREAD

**Discussion by staceywhitmore-inl** [Kitware/trame#543] [Trame/ParaView] [answered]
I've searched the discussions and can't seem to find much on this topic but what methods might I call to access camera controls such as Move Speed, Rotate Speed, and Zoom Speed in a Trame application that is remote rendered using a Paraview instance on the backend? 
So, far the only camera method I've come across on the controller is .view_reset_camera(). (...and  .CameraPosition(), .CameraFocalPoint(), and .CameraParallelScale() for the active render view). 

How might I accessing these 3 camera controls (move speed, rotate speed, and zoom speed) so that I can add VSliders to adjust them? 
Or where is a good place to search for documentation on these methods? 

Also, once I've established a desired camera angle, are there ready-baked methods for easily taking a screenshot of the model in the window to save it as a .jpg or .png and wiring this functionality to say a VButton? Or, will I need to write some custom Javascript for this or tap into an existing API such as Chrome's Broswser API for something like [captureVisibleTab](https://developer.chrome.com/docs/extensions/reference/api/tabs#method-captureVisibleTab) ?

**Comment by jourdain** [comment]
I guess the question would be how you set those in ParaView or in VTK from Python. So not a trame question.

For capturing the image you should be using [VTK/ParaView API](https://github.com/Kitware/trame/blob/master/examples/validation/core/24_vtk_download_image.py#L56-L67) and then provide a mechanism to [download that file to the client](https://kitware.github.io/trame/examples/core/files.html#download).

**Reply by staceywhitmore-inl** [reply]
I see. So, I would just adjust those camera parameters in Paraview and then run a trace to get the output. Is there Discussions board similar to this one for Paraview or VTK? If so, I looks like I posted my first question on the wrong forum. My bad. Thanks!

**Reply by jourdain** [reply]
https://discourse.paraview.org/
https://discourse.vtk.org/

**Reply by staceywhitmore-inl** [reply]
I got the [screenshot example](https://github.com/Kitware/trame/blob/master/examples/validation/core/24_vtk_download_image.py#L56-L67) to work on its own; however, since I'm doing remote rendering with Paraview on my Trame app would this example still apply as is? 

For instance, rather than using the sample vtkConeSource() when setting up the mapper's Input connection on line 35 of the example, I would be using `simple.TIFFSeriesReader()` as my source (and later extracting a couple subsets from that). 

Also, I've just been setting my self.view to `self.view = simple.Render()` ; however, this example sets up a `renderWindow` with a `vtkRenderer()` (along with a mapper and actor which I hitherto have not needed to get my script to work as they don't seem to be used much in the parview-with-remote-rendering examples). 

Does this mean that to enable screenshots I will need to switch to using a `renderWindow.Render()` call (perhaps utilizing a mapper and an actor) rather than just setting my `self.view` = to `simple.Render() and passing that to my `paraview.VtkRemoteView()` at the end?

In short, this issues I seem to be having are with passing my TiffSeriesReader()'s source's GetOutputPort()  to the `mapper.SetInputConnection()` (from line 35 of the example) as it leads to the following error 

`AttributeError: 'NoneType' object has no attribute 'IsA'
`
once it gets to the `paraview_widgets.VtkRemoteView()` portion at the end of my trame script. And omitting this section it downloads the PNG file but any image app I use to read it will complain it doesn't support the file format (since I'm probably missing the step that would allow the PNG to be formated properly and therfore readable). 

As always feel free to provide any documentation you think might point me in the right directions and I can provide a code snippet of what I'm trying to do if that helps.

**Reply by jourdain** [reply]
The main issue here is that you are comparing a VTK script vs a ParaView one. While both use VTK underneath they can not be intermixed unless you really know what you are doing and have a deep understanding of ParaView proxies. 

In other words, you need to use the Python API of ParaView to take a screenshot (which is given at the bottom of the trace as reference but pasted below for simplicity).

```python
## Save a screenshot of the active view
# SaveScreenshot("path/to/screenshot.png")
```

As you can see, that will generate a file on the server side. Therefore you will need to leverage trame API to let the user download such binary content. From the example you linked, you have that piece covered. I guess the next missing piece would be how to read a file in Python and give its bytes to trame... Well, you can use [that API](https://docs.python.org/3/library/pathlib.html#pathlib.Path.read_bytes) for it.

**Reply by staceywhitmore-inl** [reply]
I see. I added `simple.SaveScreenshot("/server/path/to/screenshot.png")` to my `download_screenshot()` trigger [referenced here on my VBtn's click attribute]:
`click="utils.download('vtk.png', trigger('download_screenshot'), 'image/png')",
`
and sure enough—there was a screenshot.png saved to that path which looked just like my screen when I opened it but the vtk.png that was also downloaded to downloads directory is still unreadable bytecode so inside my download_screenshot @trigger method I will also need to have it read that screenshot.png file's bytes in Python using something like the path.read_bytes() method mentioned in the link you provided and 'give it's bytes to trame'. 

My next question is what should I have trame do with those bytes? From the previous linked example I can see that 
`server.protocol.addAttachment(memoryview(writer1.GetResult()))
`
returns a type() string. Is it just a matter of joining that byte array `p.read_byes()` into a string (`str()?`)and returning that back to `utils.download()` from the trame API (used in the button). In short, what is trame expecting me to do those bytes I read from the .PNG saved on the server and returned by my `download_screenshot()` method?

I've added the following code to my `download_screenshot` trigger
following the suggestion in [this Stack Overflow post](https://stackoverflow.com/questions/18491416/pil-convert-bytearray-to-image) for converting 
```
    @trigger("download_screenshot") # @ctrl.trigger
    def download_screenshot(self):
        SERVER_PATH = Path("D:\\DATA\\screenshot.png")
        res = simple.SaveScreenshot("D:\\DATA\\screenshot.png")
        image_bytes = SERVER_PATH.read_bytes()
        image = Image.open(io.BytesIO(image_bytes))

        return image
```
But this would be overkill it's just expecting a large strigified version of all the bytes in the array. Regardless of whether I have my function above return an image converted by `open(io.bytes())`, a string the image bytes or just the bytes array itself,
The screenshot still saves to the 'server' (which is just my computer at this point) but I get the following error in my console 
`JS Error => [object Object]`
In JS this typically means I need to wrap whatever I'm sending in `JSON.stringify()` If this, is the issue can I make such a JavaScript call to `JSON.stringify()` from my Trame Python script? 

If the answer to my what-to-have-trame-do-the-bytes question is too much at this time at least, I'm able to get the PNG saved to the server so I will be happy with that for now. If I'm using straight JS I  knw just how to get to the client's downloads directory. I would just need to make sure to do it the Trame way as I'm less familiar with the python esp. of the trame variety. So I can worry about that detail later if needed and I can be satisfied with this for now. 
Thanks again for sharing the `SaveScreenshot()` fix. That was very helpful.

**Reply by jourdain** [reply]
```python
@trigger("download_screenshot") # @ctrl.trigger
def download_screenshot(self):
        SERVER_PATH = Path("D:\\DATA\\screenshot.png")
        simple.SaveScreenshot(str(SERVER_PATH))
        return server.protocol.addAttachment(SERVER_PATH.read_bytes())
```

**Reply by jourdain** [reply]
[This is covered](https://kitware.github.io/trame/examples/core/files.html) in the documentation.

**Reply by staceywhitmore-inl** [reply]
I'll take a look. Thanks!

**Reply by jourdain** [reply, marked as answer]
Actually with trame 3.6+ you should be able to just do 

```python
@trigger("download_screenshot") # @ctrl.trigger
def download_screenshot(self):
        SERVER_PATH = Path("D:\\DATA\\screenshot.png")
        simple.SaveScreenshot(str(SERVER_PATH))
        return SERVER_PATH.read_bytes()
```

**Reply by staceywhitmore-inl** [reply]
Yes, this worked. Thanks!

**Reply by moon-chyldren** [reply]
I’ve successfully got the save screenshot function to work before with a trame-paraview app but the trame-paraview app I’m currently working on is giving me the following error when I click the button to download a screenshot. 
```bash
ValueError: A view or layout must be specified
```
proceeded by the error @staceywhitmore-inl mentioned above 
```
JS Error => [object object]
```
My code for the trigger function looks exactly like what @staceywhitmore-inl mentioned above. 
I’m wondering if it has something to do with the way my app is laid out in my `build_ui()` function. 
For instance, when I got this to work before I was using `vtkRemoteLocalView()` but this time I’m just using `vtkRemoteView()` like so
```python 
with VAppLayout(server, full_height=True) as layout:
       client.Style(‘html { overflow-y: hidden; }’)
       with pv_widgets.vtkRemoteView(self.render_window, interactive_quality=50) as remote_view:
               ctrl.view_update = remote_view.update()
               # ...
```

Has anyone seen this ValueError before and happen to know what’s generating it? And are there other ways (or methods) for calling `simple.SaveScreenshot()` from somewhere other than `simple`? This line seems to be where the error is originating from as the console indicates it comes from `…\paraview\simiple.py”, line 1468, in SaveScreenshot` so I can look there in the meantime to see if I can figure out what’s causing it.
I can provide more details if needed.
